hack/generate/changelog/gen-changelog.go

```diff
@@ -0,0 +1,331 @@
+		return fmt.Errorf("could not write migration guide: %v", err)
+	}
+	return nil
+}
+
+type fragment struct {
+	Entries []entry `yaml:"entries"`
+}
+
+type entry struct {
+	Description string     `yaml:"description"`
+	Kind        entryKind  `yaml:"kind"`
+	Breaking    bool       `yaml:"breaking"`
+	Migration   *migration `yaml:"migration,omitempty"`
+	PullRequest *uint      `yaml:"pull_request_override,omitempty"`
+}
+
+type entryKind string
+
+const (
+	addition    entryKind = "addition"
+	change      entryKind = "change"
+	removal     entryKind = "removal"
+	deprecation entryKind = "deprecation"
+	bugfix      entryKind = "bugfix"
+)
+
+func (k entryKind) toHeader() string {
+	switch k {
+	case addition:
+		return "### Additions"
+	case change:
+		return "### Changes"
+	case removal:
+		return "### Removals"
+	case deprecation:
+		return "### Deprecations"
+	case bugfix:
+		return "### Bug Fixes"
+	default:
+		panic("invalid entry kind; NOTE TO DEVELOPERS: check entryKind.validate")
+	}
+}
+
+type migration struct {
+	Header string `yaml:"header"`
+	Body   string `yaml:"body"`
+}
+
+type config struct {
+	File    string
+	Title   string
+	Entries []entry
+}
+
+func (f *fragment) validate() error {
+	for i, e := range f.Entries {
+		if err := e.validate(); err != nil {
+			return fmt.Errorf("entry[%d] invalid: %v", i, err)
+		}
+	}
+	return nil
+}
+
+func (e *entry) validate() error {
+	if err := e.Kind.validate(); err != nil {
+		return fmt.Errorf("invalid kind: %v", err)
+	}
+
+	if e.Breaking && e.Kind != change && e.Kind != removal {
+		return fmt.Errorf("breaking changes can only be kind %q or %q, got %q", change, removal, e.Kind)
+	}
+
+	if e.Breaking && e.Migration == nil {
+		return fmt.Errorf("breaking changes require migration descriptions")
+	}
+
+	if e.Migration != nil {
+		if err := e.Migration.validate(); err != nil {
+			return fmt.Errorf("invalid migration: %v", err)
+		}
+	}
+	return nil
+}
+
+func (e entry) toChangelogString() string {
+	text := strings.Trim(e.Description, "\n")
+	if e.Breaking {
+		text = fmt.Sprintf("**Breaking change**: %s", text)
+	}
+	if !strings.HasSuffix(text, ".") && !strings.HasSuffix(text, "!") {
+		text = fmt.Sprintf("%s.", text)
+	}
+	if e.PullRequest != nil {
+		text = fmt.Sprintf("%s (%s)", text, e.pullRequestLink())
+	}
+	return fmt.Sprintf("- %s", text)
+}
+
+func (e entry) pullRequestLink() string {
+	const repo = "github.com/operator-framework/operator-sdk"
+	return fmt.Sprintf("[#%d](https://%s/pull/%d)", *e.PullRequest, repo, *e.PullRequest)
+}
+
+func getPullRequest(filename string) (uint, error) {
+	args := fmt.Sprintf("log --follow --pretty=format:%%s --diff-filter=A --find-renames=40%% %s", filename)
+	line, err := exec.Command("git", strings.Split(args, " ")...).CombinedOutput()
+	if err != nil {
+		return 0, fmt.Errorf("failed to locate git commit for PR discovery: %v", err)
+	}
+
+	numRegex := regexp.MustCompile(`\(#(\d+)\)$`)
```

Comment:
Nit: move this into a global var so it doesn't get recompiled on each call, although the go compiler might be smart enough to figure that out.